src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/perform_interaction_request.cc

```diff
@@ -1063,13 +1091,73 @@ void PerformInteractionRequest::SendBothModeResponse(
       msg_param.empty() ? NULL : &msg_param;
   std::string info = app_mngr::commands::MergeInfos(
       ui_perform_info, ui_info_, vr_perform_info, vr_info_);
+
   DisablePerformInteraction();
+
   SendResponse(result,
                perform_interaction_result_code,
                info.empty() ? NULL : info.c_str(),
                response_params);
 }
 
+mobile_apis::Result::eType
+PerformInteractionRequest::PrepareResultCodeForResponse(
+    const app_mngr::commands::ResponseInfo& ui_response,
+    const app_mngr::commands::ResponseInfo& vr_response) {
+  LOG4CXX_DEBUG(logger_,
+                "InteractionMode = " << static_cast<int32_t>(interaction_mode_)
+                                     << " | FirstAnsweredInterface = "
+                                     << static_cast<int32_t>(first_responser_));
+
+  if (mobile_apis::InteractionMode::VR_ONLY == interaction_mode_) {
+    if (FirstAnsweredInterface::VR == first_responser_) {
+      return MessageHelper::HMIToMobileResult(vr_result_code_);
+    }
+  }
+
+  if (mobile_apis::InteractionMode::BOTH == interaction_mode_) {
+    if (IsVRPerformInteractionResponseSuccessfulInBothMode()) {
+      return MessageHelper::HMIToMobileResult(vr_result_code_);
+    }
+    return MessageHelper::HMIToMobileResult(ui_result_code_);
+  }
+
+  return CommandRequestImpl::PrepareResultCodeForResponse(ui_response,
+                                                          vr_response);
+}
+
+bool PerformInteractionRequest::PrepareResultForMobileResponse(
+    app_mngr::commands::ResponseInfo& ui_response,
+    app_mngr::commands::ResponseInfo& vr_response) const {
+  if (mobile_apis::InteractionMode::VR_ONLY == interaction_mode_) {
+    if (FirstAnsweredInterface::VR == first_responser_) {
+      return vr_response.is_ok;
+    }
+  }
+
+  if (mobile_apis::InteractionMode::BOTH == interaction_mode_) {
+    return (vr_response.is_ok || ui_response.is_ok);
+  }
+
+  return CommandRequestImpl::PrepareResultForMobileResponse(ui_response,
+                                                            vr_response);
+}
+
+bool PerformInteractionRequest::
+    IsVRPerformInteractionResponseSuccessfulInBothMode() {
+  using namespace mobile_apis;
+  app_mngr::commands::ResponseInfo vr_perform_info(
+      vr_result_code_, HmiInterfaces::HMI_INTERFACE_VR, application_manager_);
+  return (vr_perform_info.is_ok && InteractionMode::BOTH == interaction_mode_);
+}
+
+void PerformInteractionRequest::StoreFirstAnsweredInterface(
```

Comment:
unnecessary function, please remove

Reply:
@iCollin We need this function to save the first answered interface and use the saved value for the case when the current interaction mode is `InteractionMode::VR_ONLY` and SDL received the `VR_PerformInteraction` response. In this case, SDL Core should send the `UI_ClosePopUp` request to the HMI to close the UI menu.

Reply:
@IGapchuk sorry, I should have explained more. It is a useful function, but since it will only ever be used in two specific places it actually takes more code to define the function than to just write the function body inline.